naga/src/front/wgsl/lower/conversion.rs

```diff
@@ -0,0 +1,378 @@
+        Some((expr_scalar, goal_scalar))
+    }
+
+    fn automatically_convertable_scalar(
+        &self,
+        types: &crate::UniqueArena<crate::Type>,
+    ) -> Option<crate::Scalar> {
+        use crate::TypeInner as Ti;
+        match *self {
+            Ti::Scalar(scalar) | Ti::Vector { scalar, .. } | Ti::Matrix { scalar, .. } => {
+                Some(scalar)
+            }
+            Ti::Array { base, .. } => types[base].inner.automatically_convertable_scalar(types),
+            Ti::Atomic(_)
+            | Ti::Pointer { .. }
+            | Ti::ValuePointer { .. }
+            | Ti::Struct { .. }
+            | Ti::Image { .. }
+            | Ti::Sampler { .. }
+            | Ti::AccelerationStructure
+            | Ti::RayQuery
+            | Ti::BindingArray { .. } => None,
+        }
+    }
+}
+
+impl crate::Scalar {
+    /// Find the common type of `self` and `other` under WGSL's
+    /// automatic conversions.
+    ///
+    /// If there are any scalars to which WGSL's automatic conversions
+    /// will convert both `self` and `other`, return the best such
+    /// scalar. Otherwise, return `None`.
+    pub const fn automatic_conversion_join(self, other: Self) -> Option<crate::Scalar> {
+        use crate::ScalarKind as Sk;
+
+        match (self.kind, other.kind) {
+            // When the kinds match...
+            (Sk::AbstractFloat, Sk::AbstractFloat)
+            | (Sk::AbstractInt, Sk::AbstractInt)
+            | (Sk::Sint, Sk::Sint)
+            | (Sk::Uint, Sk::Uint)
+            | (Sk::Float, Sk::Float)
+            | (Sk::Bool, Sk::Bool) => {
+                if self.width == other.width {
+                    // ... either no conversion is necessary ...
+                    Some(self)
+                } else {
+                    // ... or no conversion is possible.
+                    // We never convert concrete to concrete, and
+                    // abstract types should have only one size.
+                    None
+                }
+            }
+
+            // AbstractInt converts to AbstractFloat.
+            (Sk::AbstractFloat | Sk::AbstractInt, Sk::AbstractFloat | Sk::AbstractInt) => {
+                Some(Self {
+                    kind: Sk::AbstractFloat,
+                    width: crate::ABSTRACT_WIDTH,
+                })
+            }
+
+            // AbstractFloat converts to Float.
+            (Sk::AbstractFloat, Sk::Float) => Some(Self::float(other.width)),
+            (Sk::Float, Sk::AbstractFloat) => Some(Self::float(self.width)),
+
+            // AbstractInt converts to concrete integer or float.
+            (Sk::AbstractInt, kind @ (Sk::Uint | Sk::Sint | Sk::Float)) => Some(Self {
+                kind,
+                width: other.width,
+            }),
+            (kind @ (Sk::Uint | Sk::Sint | Sk::Float), Sk::AbstractInt) => Some(Self {
+                kind,
+                width: self.width,
+            }),
```

Comment:
```suggestion
            (Sk::AbstractFloat, Sk::AbstractInt) => Some(other),
            (Sk::AbstractInt, Sk::AbstractFloat) => Some(self),

            // AbstractFloat converts to Float.
            (Sk::AbstractFloat, Sk::Float) => Some(other),
            (Sk::Float, Sk::AbstractFloat) => Some(self),

            // AbstractInt converts to concrete integer or float.
            (Sk::AbstractInt, Sk::Uint | Sk::Sint | Sk::Float) => Some(other),
            (Sk::Uint | Sk::Sint | Sk::Float, Sk::AbstractInt) => Some(self),
```

I find this easier to follow.

Reply:
Very much so.